plugins/sentinelone/komand_sentinelone/tasks/monitor_activities_and_events/task.py

```diff
@@ -0,0 +1,327 @@
+EVENTS_LAST_LOG_TIMESTAMP = "events_last_log_timestamp"
+THREATS_LAST_LOG_TIMESTAMP = "threats_last_log_timestamp"
+ACTIVITIES_PAGE_CURSOR = "activities_page_cursor"
+EVENTS_PAGE_CURSOR = "events_page_cursor"
+THREATS_PAGE_CURSOR = "threats_page_cursor"
+
+# Cutoff values
+DATE_TIME_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"
+INITIAL_CUTOFF_HOURS = 24
+MAX_CUTOFF_HOURS = 24 * 7
+LOGS_PAGE_LIMIT = 1000
+
+API_VERSION = "2.1"
+STARTING_TIMESTAMP = "1900-01-01T00:00:00.000000Z"
+
+ACTIVITIES_LOGS = "activities_logs"
+EVENTS_LOGS = "events_logs"
+THREATS_LOGS = "threats_logs"
+
+LOG_TYPE_MAP = {
+    ACTIVITIES_LOGS: {"LAST_LOG_TIMESTAMP": ACTIVITIES_LAST_LOG_TIMESTAMP, "PAGE_TOKEN": ACTIVITIES_PAGE_CURSOR},
+    EVENTS_LOGS: {"LAST_LOG_TIMESTAMP": EVENTS_LAST_LOG_TIMESTAMP, "PAGE_TOKEN": EVENTS_PAGE_CURSOR},
+    THREATS_LOGS: {"LAST_LOG_TIMESTAMP": THREATS_LAST_LOG_TIMESTAMP, "PAGE_TOKEN": THREATS_PAGE_CURSOR},
+}
+
+
+class MonitorActivitiesAndEvents(insightconnect_plugin_runtime.Task):
+    def __init__(self):
+        super(self.__class__, self).__init__(
+            name="monitor_activities_and_events",
+            description=Component.DESCRIPTION,
+            input=MonitorActivitiesAndEventsInput(),
+            output=MonitorActivitiesAndEventsOutput(),
+            state=MonitorActivitiesAndEventsState(),
+        )
+
+    def run(self, params={}, state={}, custom_config={}):
+        """
+        Query activities, device control events, and threats logs between a supplied timeframe and the current time.
+        Return all found logs and set last task run time, pagination cursors, and last log timestamps for each log type.
+        """
+        existing_state = state.copy()
+        try:
+            is_initial_run = self.check_initial_run(state)
+            queries = self.check_queries(params)
+            cursors = self.get_cursors(state)
+            total_queries = self.count_queries(queries)
+            total_forbidden_responses = 0
+            last_run_timestamp = state.get(LAST_RUN_TIMESTAMP)
+            self.logger.info(cursors)
+            is_paginating = any(cursors.values())
+            if is_paginating:
+                self.log_pagination_cycle(cursors)
+
+            current_run_timestamp, lookback_timestamp = self.get_lookback_values(
+                is_initial_run, custom_config, last_run_timestamp
+            )
+
+            all_logs, total_forbidden_responses = self.collect_logs_handler(
+                queries,
+                cursors,
+                state,
+                lookback_timestamp,
+                current_run_timestamp,
+                total_forbidden_responses,
+                is_paginating,
+            )
+
+            # Check if all ran queries have returned 401 or 403 errors and raise an exception if so
+            if total_forbidden_responses >= total_queries > 0:
+                return [], existing_state, False, 401, PluginException(preset=PluginException.Preset.UNAUTHORIZED)
```

Comment:
Is it worth adding a log line here and for the exceptions below to make troubleshooting easier if these exceptions are thrown?